snowfall/data/asr_datamodule.py

```diff
@@ -0,0 +1,211 @@
+from pathlib import Path
+
+import argparse
+
+from typing import List, Union
+
+import logging
+
+from torch.utils.data import DataLoader
+
+from lhotse import Fbank, FbankConfig, load_manifest
+from lhotse.dataset import BucketingSampler, CutConcatenate, CutMix, K2SpeechRecognitionDataset, SingleCutSampler, \
+    SpecAugment
+from lhotse.dataset.input_strategies import OnTheFlyFeatures
+from snowfall.common import str2bool
+from snowfall.data.datamodule import DataModule
+
+
+class AsrDataModule(DataModule):
+    """
+    DataModule for K2 ASR experiments.
+    It assumes there is always one train and valid dataloader,
+    but there can be multiple test dataloaders (e.g. LibriSpeech test-clean and test-other).
+
+    It contains all the common data pipeline modules used in ASR experiments, e.g.:
+    - dynamic batch size,
+    - bucketing samplers,
+    - cut concatenation,
+    - augmentation,
+    - on-the-fly feature extraction
+
+    This class should be derived for specific corpora used in ASR tasks.
+    """
+
+    @classmethod
+    def add_arguments(cls, parser: argparse.ArgumentParser):
+        super().add_arguments(parser)
+        group = parser.add_argument_group(
+            title='ASR data related options',
+            description='These options are used for the preparation of PyTorch DataLoaders '
+                        'from Lhotse CutSet\'s -- they control the effective batch sizes, '
+                        'sampling strategies, applied data augmentations, etc.'
+        )
+        group.add_argument(
+            '--feature-dir',
+            type=Path,
+            default=Path('exp/data'),
+            help='Path to directory with train/valid/test cuts.'
+        )
+        group.add_argument(
+            '--max-duration',
+            type=int,
+            default=500.0,
+            help="Maximum pooled recordings duration (seconds) in a single batch.")
+        group.add_argument(
+            '--bucketing-sampler',
+            type=str2bool,
+            default=False,
+            help='When enabled, the batches will come from buckets of '
+                 'similar duration (saves padding frames).')
+        group.add_argument(
+            '--num-buckets',
+            type=int,
+            default=30,
+            help='The number of buckets for the BucketingSampler'
+                 '(you might want to increase it for larger datasets).')
+        group.add_argument(
+            '--concatenate-cuts',
+            type=str2bool,
+            default=True,
+            help='When enabled, utterances (cuts) will be concatenated '
```

Comment:
I am worried about the potential issue of the self-attention in the encoder part of the transformer/conformer model if different waves are concatenated.

It will definitely happen that frames in one wave can attend to frames from another wave
if concatenation is used since the encoder part does not use **masked** self-attention.

@danpovey @pzelasko  What do you think?

Reply:
I think it’s worth checking the model performance by training with concatenation disabled. If the WER is better then you’re right, if the WER is worse then concatenation could act as a regularizer (the model has to learn to attend to the right frames). It could also have no effect, but I don’t think we know which is true as of now.

Reply:
My recollection is that we tested this and it was OK, but we should test again.  I believe the internal attention code may support a quadratic mask that tells you whether a particular frame can attend to a particular other frame;  in principle we could pass a mask of that type in, although it would change interfaces.